### PR TITLE
Blocchi opzionali innestati

### PR DESCRIPTION
**Pull request by konova8**
- [x] FSA nel README
- [x] Creare i test
- [x] Aggiornare l'automa
- [x] Aggiornare le regex
- [x] Aggiornare json
- [x] Aggiornare json-schema
- [x] Spostare l'FSA sulla wiki

### CODE REVIEW

**Comment by konova8** [comment]
In serata provo di fare le ultime due cose nella lista, se intanto vedete cose segnalatemele qui

**Comment by lucat1** [comment]
Tocca aggiornare anche la BNF

**Comment by konova8** [comment]
C'e' un qualche problema in `parser.go` per cui non riesce ad aggiungere alla struttura dati i vari segmenti se ci sono dei segmenti nestati, ma non capisco perche' non funzioni

Se vuoi darci un'occhiata @lucat1 intanto, io riprendo a lavorarci stasera o domani

**Comment by lucat1** [comment]
bisognerà aggiornare anche il json e il suo relativo json schema

File: parser.go [hunk 1]
```diff
@@ -132,8 +132,12 @@ func clear(seg *Segment) {
 	seg.Optional = false
 }
 
-func push(segments []Segment, seg *Segment) (updatedSegments []Segment) {
-	updatedSegments = append(segments, *seg)
+func push(segments []Segment, seg *Segment, nested bool) (updatedSegments []Segment) {
+	if nested {
+		updatedSegments = append(segments[len(segments)-1].NestedSeg, *seg)
```
Comment (lucat1) [inline parser.go:137]: Devi assegnare a `segments[len(segments)-1].NestedSeg`

**Review by lucat1 — DISMISSED** [review]

File: parser.go (hunk 1 above)
Comment (konova8) [reply parser.go:137]: Non viene restituito cosi' pero', perche' noi ritorniamo `updatedSegments`

Inoltre sto cercando un'implementazione che permette annidamento infinito, ora impossibile

Comment (lucat1) [reply parser.go:137]: Tocca usare i puntatori allora.

**Comment by foxyseta** [comment]
@konova8 ho risolto un conflitto che si è creato oggi traslocando la documentazione sulla wiki. Ora puoi continuare a lavorare su questa PR senza problemi. Se vuoi apporre modifiche alla documentazione puoi farlo sempre dalla wiki (la pagina esatta è segnalata nel README.me di questa repository).

**Comment by konova8** [comment]
> @konova8 ho risolto un conflitto che si è creato oggi traslocando la documentazione sulla wiki. Ora puoi continuare a lavorare su questa PR senza problemi. Se vuoi apporre modifiche alla documentazione puoi farlo sempre dalla wiki (la pagina esatta è segnalata nel README.me di questa repository).

Ho visto, avevo aggiunto un elemento alla lista di cose da fare nel primo messaggio della PR. Hai fatto prima tu però, grazie!

**Comment by konova8** [comment]
Dovrebbe essere funzionante ora, ho aggiunto qualche test e mi sembra sia tutto in ordine.
Non abbiamo considerato che `Format` possa dare errore e l'ho lasciata così, tanto da cli prima facciamo il parsing di un file `.synta` quindi al massimo l'errore lo abbiamo lì, corretto?

**Comment by foxyseta** [comment]
@lucat1 scusa potresti fare tu la recensione?

**Comment by foxyseta** [comment]
@lucat1 bump

**Review by lucat1 — APPROVED** [review]
Ci ho dato un occhio e mi pare tutto sensato.
Non ho guardato al codice dell'automa, ma se passano quei test assumo che sia funzionante. Magari qualche edge case non è coperto, ma non ho tempo per controllare ore.

Per me può essere approvata. Se si trovano dei bug in qualche edge case sarebbe veloce sistemare direi.
